src/Microsoft.AspNet.JsonPatch/Adapters/ObjectAdapter.cs

```diff
@@ -133,91 +126,242 @@ public void Add([NotNull] Operation operation, [NotNull] object objectToApplyTo)
-                                operationToReport,
-                                Resources.FormatInvalidIndexForArrayProperty(operationToReport.op, path)));
+                            // get the actual type
+                            var typeOfPathProperty = treeAnalysisResult.Container
+                                .GetValueForCaseInsensitiveKey(treeAnalysisResult.PropertyPathInParent).GetType();
 
-                            return;
+                            // can the value be converted to the actual type?
+                            var conversionResultTuple =
+                                ConvertToActualType(typeOfPathProperty, value);
+
+                            // conversion successful
+                            if (conversionResultTuple.CanBeConverted)
+                            {
+                                treeAnalysisResult.Container.SetValueForCaseInsensitiveKey(treeAnalysisResult.PropertyPathInParent,
+                                    conversionResultTuple.ConvertedInstance);
+                            }
+                            else
+                            {
+                                LogError(new JsonPatchError(
+                                   objectToApplyTo,
+                                   operationToReport,
+                                   Resources.FormatInvalidValueForProperty(conversionResultTuple.ConvertedInstance, path)));
+                            }
                         }
                     }
+                    else
+                    {
+                        // New property - add it.  
+                        treeAnalysisResult.Container.Add(treeAnalysisResult.PropertyPathInParent, value);
+                    }
                 }
                 else
                 {
                     LogError(new JsonPatchError(
-                        objectToApplyTo,
-                        operationToReport,
-                        Resources.FormatInvalidIndexForArrayProperty(operationToReport.op, path)));
-
-                    return;
+                                  objectToApplyTo,
+                                  operationToReport,
+                                  Resources.FormatPropertyCannotBeAdded(path)));
                 }
             }
             else
             {
-                var conversionResultTuple = ConvertToActualType(
-                    patchProperty.Property.PropertyType,
-                    value);
+                if (!treeAnalysisResult.IsValidPathForAdd)
+                {
+                    LogError(new JsonPatchError(
+                                     objectToApplyTo,
+                                     operationToReport,
+                                     Resources.FormatPropertyDoesNotExist(path)));
+                }
+
+                // If it' an array, add to that array.  If it's not, we replace.
+
+                // is the path an array (but not a string (= char[]))?  In this case,
```

Comment:
Treating string as an array of char would make path "person/name" invalid - will clarify it in the comments